Sources/SwiftLexicalLookup/ScopeImplementations.swift

```diff
@@ -0,0 +1,289 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftSyntax
+
+extension SyntaxProtocol {
+  /// Parent scope of this syntax node, or scope introduced by this syntax node.
+  var scope: ScopeSyntax? {
+    if let scopeSyntax = Syntax(self).asProtocol(SyntaxProtocol.self) as? ScopeSyntax {
+      scopeSyntax
+    } else {
+      self.parent?.scope
+    }
+  }
+}
+
+@_spi(Experimental) extension SourceFileSyntax: ScopeSyntax {
+  /// All names introduced in the file scope
+  /// according to the default strategy: `memberBlockUpToLastDecl`.
+  public var introducedNames: [LookupName] {
+    introducedNames(using: .memberBlockUpToLastDecl)
+  }
+
+  /// All names introduced in the file scope
+  /// using the provided configuration.
+  ///
+  /// Example usage:
+  /// ```swift
+  /// class a {}
+  /// class b {
+  ///   // <--
+  /// }
+  /// let c = 0
+  /// class d {}
+  /// if true {}
+  /// class e {}
+  /// let f = 0
+  /// ```
+  /// During lookup, according to different configurations,
+  /// names available at the marked place are:
+  /// - for `fileScopeNameIntroductionStrategy` - a, b, c, d
+  /// - for `memberBlock` - a, b, c, d, e, f
+  /// - for `codeBlock` - a
+  public func introducedNames(using fileScopeHandling: FileScopeHandlingConfig) -> [LookupName] {
+    switch fileScopeHandling {
+    case .memberBlockUpToLastDecl:
+      var encounteredNonDeclaration = false
+
+      return statements.flatMap { codeBlockItem in
+        let item = codeBlockItem.item
+
+        if encounteredNonDeclaration {
+          return LookupName.getNames(from: item, accessibleAfter: codeBlockItem.endPosition)
+        } else {
+          if item.is(DeclSyntax.self) || item.is(VariableDeclSyntax.self) {
+            return LookupName.getNames(from: item)
+          } else {
+            encounteredNonDeclaration = true
+            return LookupName.getNames(from: item, accessibleAfter: codeBlockItem.endPosition)
```

Comment:
Those should be all names introduced before and including the non-declaration. In this case:
```swift
struct a {} // <-- in result

let a = 0 // <-- in result

if a == 0 {}

a // <-- lookup from here

class a {}
```
The lookup result would consist of `struct a` and `let a = 0`, not including the bottommost `class a`.

Reply:
I guess `a` is still an invalid redeclaration but name lookup still works. That makes sense.

Does name lookup work for the following?

```swift
func foo() {
  let a = A() // should refer to `A` declared later.
  class A {}
}
```

Reply:
It doesn't work indeed. The class declaration behaves like a typical variable declaration. I think it could probably be easily solved with the name introduction rework I thought about and mentioned in the other comment. By moving the logic to syntax nodes we could greatly customize it for each node.

Reply:
👍🏽 Would be good to see this fixed in a follow-up PR.

Reply:
I added it to #2755 with an [appropriate test case](https://github.com/swiftlang/swift-syntax/blob/e1831bee74bd545421252fffb906e37438da2809/Tests/SwiftLexicalLookupTest/NameLookupTests.swift#L567-L583).